semantic_release/hvcs/bitbucket.py

```diff
@@ -0,0 +1,125 @@
+# incorrect values, which causes TensorBoard to fail to render on
+# the frontend.
+# This method hard-codes the correct mappings for certain MIME
+# types that are known to be either used by python-semantic-release or
+# problematic in general.
+mimetypes.add_type("application/octet-stream", ".whl")
+mimetypes.add_type("text/markdown", ".md")
+
+
+class Bitbucket(HvcsBase):
+    """Bitbucket helper class"""
+
+    API_VERSION = "2.0"
+    DEFAULT_DOMAIN = "bitbucket.org"
+    DEFAULT_API_DOMAIN = "api.bitbucket.org"
+    DEFAULT_ENV_TOKEN_NAME = "BITBUCKET_TOKEN"
+
+    def __init__(
+        self,
+        remote_url: str,
+        hvcs_domain: str | None = None,
+        hvcs_api_domain: str | None = None,
+        token: str | None = None,
+    ) -> None:
+        self._remote_url = remote_url
+
+        if hvcs_domain is not None:
+            self.hvcs_domain = hvcs_domain
+        else:
+            api_url = os.getenv("BITBUCKET_SERVER_URL", self.DEFAULT_DOMAIN)
+            self.hvcs_domain = api_url.replace("https://", "")
+
+        # ref: https://developer.atlassian.com/cloud/bitbucket/rest/intro/#uri-uuid
+        if hvcs_api_domain is not None:
+            self.hvcs_api_domain = hvcs_api_domain
+        else:
+            api_url = os.getenv("BITBUCKET_API_URL", self.DEFAULT_API_DOMAIN)
+            self.hvcs_api_domain = api_url.replace("https://", "")
+
+        self.api_url = f"https://{self.hvcs_api_domain}/{self.API_VERSION}"
+
+        self.token = token
+        auth = None if not self.token else TokenAuth(self.token)
+        self.session = build_requests_session(auth=auth)
+
+    @lru_cache(maxsize=1)
+    def _get_repository_owner_and_name(self) -> tuple[str, str]:
+        # ref: https://support.atlassian.com/bitbucket-cloud/docs/variables-and-secrets/
+        if "BITBUCKET_REPO_FULL_NAME" in os.environ:
+            log.info("Getting repository owner and name from environment variables.")
+            owner, name = os.environ["BITBUCKET_REPO_FULL_NAME"].rsplit("/", 1)
+            return owner, name
+        return super()._get_repository_owner_and_name()
+
+    def compare_url(self, from_rev: str, to_rev: str) -> str:
+        """
+        Get the Bitbucket comparison link between two version tags.
+        :param from_rev: The older version to compare.
+        :param to_rev: The newer version to compare.
+        :return: Link to view a comparison between the two versions.
+        """
+        return (
+            f"https://{self.hvcs_domain}/{self.owner}/{self.repo_name}/"
+            f"branches/compare/{from_rev}%0D{to_rev}"
+        )
+
+    def remote_url(self, use_token: bool = True) -> str:
+        if not use_token:
+            # Note: Assume the user is using SSH.
+            return self._remote_url
+        if not self.token:
+            raise ValueError("Requested to use token but no token set.")
+        user = os.environ.get("BITBUCKET_USER")
```

Comment:
Thank you so much for your help with all of this. This is my first real contribution to an open source project, so I really appreciate your detailed explanations, very kind!
I added a section to the `docs/index.rst` to document the only extra environment variable for bitbucket: `BITBUCKET_USER`. Maybe it's not the best solution but it's something until we come up with anything better.

Reply:
Congratulations! It's a great feeling and I'm glad you made the endeavor. Glad I could help, and the effort you put forth in adding testing and docs is much more than most do, so keep it up!

Reply:
I have to echo the sentiment from @codejedi365 - as far as contributions go this one is really excellent @beatreichenbach, particularly for a first-time contribution too!

We will eventually have to do something about the growing number of CI systems and Git remotes. Eventually it would be really nice to rekindle the talk of plugin architectures...